Fixed #35530 -- Deprecated request.user fallback in auth.login.

jaap3 opened the pull request:
# Trac ticket number

ticket-35530

# Branch description

Removed `request.user` fallback logic in `auth.login`. There is no clear reason for it to exist, hasn't been tested until recently (through the async wraper) and is confusing at best.

# Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

felixxm commented on tests/auth_tests/test_login_logout.py:
```diff
@@ -0,0 +1,100 @@
+
+        self.assertNotIn(auth.SESSION_KEY, self.request.session)
+
+        with self.assertWarnsMessage(
+            RemovedInDjango60Warning,
+            "Fallback to request.user when user is None will be removed.",
+        ):
```
```suggestion
        msg = "Fallback to request.user when user is None will be removed."
        with self.assertWarnsMessage(RemovedInDjango60Warning, msg):
```

jaap3 commented:
Thanks Mariusz, does that mean that this is acceptable? Should I add the deprecation to the release notes?

nessita commented on django/contrib/auth/__init__.py:
```diff
@@ -11,6 +12,7 @@
 from django.utils.module_loading import import_string
 from django.views.decorators.debug import sensitive_variables
 
+from ...utils.deprecation import RemovedInDjango60Warning
```
Could you please use the full path as you did in the test file?

nessita replied:
Also: deprecation landing in 5.2 will be removed in 6.1 (not in 6.0), and we need to add docs for this. Specifically, we need to add a release note and an entry in the `docs/internals/deprecation.txt` file.

nessita commented on django/contrib/auth/__init__.py:
```diff
@@ -108,6 +110,13 @@ def login(request, user, backend=None):
     session_auth_hash = ""
     if user is None:
         user = request.user
+        if user and isinstance(user, get_user_model()):
```
We would need to add some extra commentary around the code that is being deprecated. The specifics are described in [these docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature).

nessita reviewed (changes requested):
Hello @jaap3! Thank you for your work on this :star2: 

I have added inline comments but also we should have two distinct commits as follows:

1. The adding of missing tests, without any logic change.
2. The deprecation inclusion with its tests, release notes and deprecation notes.

jaap3 commented:
Thanks @nessita, I'm a bit busy at the moment, have to find some time to get to this

nessita commented:
> Thanks @nessita, I'm a bit busy at the moment, have to find some time to get to this

No problem @jaap3, thank all the time you need. When this is ready for re-review, just update flags in the Trac ticket. Thank you!

jaap3 commented:
@nessita I've addressed your comments. Hope I did it right.

sarahboyce commented on django/contrib/auth/__init__.py:
```diff
@@ -189,7 +202,17 @@ async def alogin(request, user, backend=None):
     """See login()."""
     session_auth_hash = ""
     if user is None:
+        # Fallback to request.user when user is None.
+        # This is undocumented behaviour and will be removed in Django 6.1.
         user = await request.auser()
+        if user and isinstance(user, get_user_model()):
+            # Avoid a warning when the user remains None or becomes AnonymousUser.
+            # Those cases result in an error down the line anyway.
+            warnings.warn(
+                "Fallback to request.user when user is None will be removed.",
+                RemovedInDjango61Warning,
+                stacklevel=2,
+            )
```
```suggestion
    # RemovedInDjango61Warning.
    if user is None:
        warnings.warn(
            "Fallback to request.user when user is None will be removed.",
            RemovedInDjango61Warning,
            stacklevel=2,
        )
        user = await request.auser()
```
I think I expect something like this with the `RemovedInDjango61Warning` comment (similar for the sync case)

sarahboyce commented on docs/releases/5.2.txt:
```diff
@@ -387,3 +387,6 @@ Miscellaneous
   function is deprecated in favor of the ``find_all`` argument.
+
+* The fallback to ``request.user`` when ``user`` is ``None`` in
+  ``django.contrib.auth.login`` will be removed.
```
```suggestion
  ``django.contrib.auth.login`` and ``django.contrib.auth.alogin`` will be
  removed.
```

sarahboyce commented on docs/internals/deprecation.txt:
```diff
@@ -18,6 +18,9 @@ details on these changes.
   will be removed.
 
+* The fallback to ``request.user`` when ``user`` is ``None`` in
+  ``django.contrib.auth.login`` will be removed.
```
```suggestion
  ``django.contrib.auth.login`` and ``django.contrib.auth.alogin`` will be
  removed.
```

sarahboyce commented on tests/async/test_async_auth.py:
```diff
@@ -60,15 +61,52 @@ async def test_alogin_new_user(self):
+    async def test_alogin_without_user_no_request_user(self):
+        request = HttpRequest()
+        request.session = await self.client.asession()
+        # RemovedInDjango61Warning: Fallback to request.user will be removed.
+        # This will raise AttributeError: 'NoneType' object has no attribute '_meta'
+        # when the fallback is removed.
```
```suggestion
        # RemovedInDjango61Warning: When the deprecation ends, replace with:
        # with self.assertRaisesMessage(
        #     AttributeError,
        #     "'NoneType' object has no attribute '_meta'",
        # ):
        #     await alogin(request, None)
```

sarahboyce commented on tests/async/test_async_auth.py:
```diff
@@ -60,15 +61,52 @@ async def test_alogin_new_user(self):
         self.assertEqual(user.username, second_user.username)
 
-    async def test_alogin_without_user(self):
+    async def test_alogin_without_user_no_request_user(self):
```
```suggestion
    # RemovedInDjango61Warning: When the deprecation ends, replace with:
    # async def test_alogin_without_user(self):
    async def test_alogin_without_user_no_request_user(self):
```
I think when we get rid of the code we only need one test for the case `user=None`

sarahboyce commented on tests/async/test_async_auth.py:
```diff
@@ -60,15 +61,52 @@ async def test_alogin_new_user(self):
+        ):
+            await alogin(request, None)
+
+    async def test_alogin_without_user_anonymous_request(self):
```
```suggestion
    # RemovedInDjango61Warning: When the deprecation ends, remove completely.
    async def test_alogin_without_user_anonymous_request(self):
```
I think we should remove this test case when we deprecate

sarahboyce commented on tests/async/test_async_auth.py:
```diff
@@ -60,15 +61,52 @@ async def test_alogin_new_user(self):
+        request.user = AnonymousUser()
+        request.auser = auser
+        request.session = await self.client.asession()
+        # RemovedInDjango61Warning: Fallback to request.user will be removed.
+        # This will raise AttributeError: 'NoneType' object has no attribute '_meta'
+        # when the fallback is removed.
```
If we plan to remove this test case, this comment is not required, otherwise should be:
```
        # RemovedInDjango61Warning: When the deprecation ends, replace with:
        # with self.assertRaisesMessage(
        #     AttributeError,
        #     "'NoneType' object has no attribute '_meta'",
        # ):
        #     await alogin(request, None)
```

sarahboyce commented on tests/async/test_async_auth.py:
```diff
@@ -60,15 +61,52 @@ async def test_alogin_new_user(self):
+            RemovedInDjango61Warning,
+            "Fallback to request.user when user is None will be removed.",
+        ):
+            # This will raise AttributeError: 'NoneType' object has no attribute '_meta'
+            # when the fallback is removed.
```
```suggestion
```
This is not needed, at the end of the deprecation we will remove this test case

sarahboyce commented on tests/auth_tests/test_login_logout.py:
I was wondering why it looks like there are no tests for login
I guess it's been tested via `self.client._login`
We could add a couple of test cases to `tests.auth_tests.test_auth_backends.BaseBackendTest` instead of here but not sure what's best

sarahboyce replied:
I also checked the current coverage level, and we don't gain much coverage from these extra tests: https://djangoci.com/job/django-coverage/HTML_20Coverage_20Report/
We appear to have strangely named test files because this is a very old app and so test files like `test_basic.py` exist
Reducing this just to the tests for the deprecation and putting them "somewhere" might be the easiest way forward here - probably need a general cleanup of the auth tests

jaap3 replied:
@sarahboyce I've now removed the logout tests as I agree those are out of scope. I renamed the file to `test_login` and kept the `TestLogin` tests intact so some sensible test module remains after deprecation. I hope that counts as putting the tests "somewhere".

sarahboyce replied:
I think this makes sense :+1: thank you

jaap3 commented:
@sarahboyce thanks for the review, I've addressed your comments. I've chosen to keep the new test module since your comment was inconclusive and I don't think these tests belong in the auth backends test suite.

sarahboyce commented on django/contrib/auth/__init__.py:
```diff
@@ -155,8 +157,15 @@ def login(request, user, backend=None):
+            RemovedInDjango61Warning,
+            stacklevel=2,
+        )
+
```
```suggestion
    # RemovedInDjango61Warning: When the deprecation ends, replace with:
    # session_auth_hash = user.get_session_auth_hash()
```
I believe this means we will also lose the guard on whether `get_session_auth_hash` exists as we are making user a required argument and it should be a user

sarahboyce commented on django/contrib/auth/__init__.py:
```diff
@@ -188,7 +197,13 @@ def login(request, user, backend=None):
+            stacklevel=2,
+        )
         user = await request.auser()
     if hasattr(user, "get_session_auth_hash"):
```
```suggestion
    # RemovedInDjango61Warning: When the deprecation ends, replace with:
    # session_auth_hash = user.get_session_auth_hash()
    if hasattr(user, "get_session_auth_hash"):
```

sarahboyce commented on docs/releases/5.2.txt:
```diff
@@ -418,3 +418,7 @@ Miscellaneous
   function is deprecated in favor of the ``find_all`` argument.
+
+* The fallback to ``request.user`` when ``user`` is ``None`` in
+  ``django.contrib.auth.login`` and ``django.contrib.auth.alogin``
+  will be removed.
```
```suggestion
  ``django.contrib.auth.login`` and ``django.contrib.auth.alogin`` will be
  removed.
```
(nit) wrap docs at 80 characters

sarahboyce commented on docs/internals/deprecation.txt:
```diff
@@ -18,6 +18,10 @@ details on these changes.
   will be removed.
 
+* The fallback to ``request.user`` when ``user`` is ``None`` in
+  ``django.contrib.auth.login`` and ``django.contrib.auth.alogin``
+  will be removed.
```
```suggestion
  ``django.contrib.auth.login`` and ``django.contrib.auth.alogin`` will be
  removed.
```
(nit) wrap docs at 80 characters

sarahboyce reviewed (approved):
Thank you @jaap3 :star: 
I pushed minor tweaks, this looks good to me :+1:
